## Lab: SQL Practice

> Author: Matt Brems

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [1]:
#!conda install psycopg2

In [3]:
# Import Pandas and Create_Engine

import pandas as pd
from sqlalchemy import create_engine

# Set up the engine to access the data.
engine = create_engine('postgres://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [4]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT count(store)
FROM stores
"""

pd.read_sql_query(sql, engine)

count
0   1973

**Prompt 1 Answer**: There are 1973 rows.

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [6]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT *
FROM stores
ORDER BY name ASC
LIMIT 4
"""

pd.read_sql_query(sql, engine)

store                                name store_status  \
0   4986  218 Fuel Express & Chubby's Liquor            A   
1   3047          3047 Cub Foods / Iowa City            I   
2   3060               3060 Cub Foods / Ames            I   
3   3061         3061 Cub Foods / Sioux City            I   

                                       store_address address_info  
0  68 Monroe St\nFloyd, IA 50435\n(43.12249790500...         None  
1  855 Highway 1 West\nIowa City, IA 522460000\n(...         None  
2  3121 Grand Ave\nAmes, IA 500100000\n(42.053458...         None  
3  1732 Hamilton Blvd\nSioux City, IA 511030000\n...         None

**Prompt 2 Answer**: 3061 Cub Foods / Sioux City

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [8]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT btl_price, bottle_qty
FROM sales
WHERE category_name = 'IMPORTED VODKA'
ORDER BY vendor
Limit 3
"""

pd.read_sql_query(sql, engine).head(3)


# SELECT f.name AS food, r.name
# FROM foods f
# INNER JOIN restaurants r ON r.id = f.restaurant_id

btl_price  bottle_qty
0     $8.99          12
1    $28.50           1
2    $28.50           6

**Prompt 3 Answer**:      
  
1. ($8.99$ ,12)  
2. ($28.50$, 1)  
3. ($28.50$, 6)    


### Prompt 4

What is the total amount of liquor sold in February 2015?

In [9]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT sum(total)
FROM sales
WHERE date >= '2015-02-01'
And date < '2015-03-01'
"""
pd.read_sql_query(sql, engine)

sum
0  21295350.55

**Prompt 4 Answer**: Total Sales after Feb 1 2015 (there are no other sales after Feb): 21,295,350.55

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [13]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT vendor, count(total) as total, SUM(bottle_qty) as qty, SUM(total)
FROM sales
WHERE category_name = 'IMPORTED VODKA'
GROUP BY vendor
ORDER BY total DESC
LIMIT 1
"""

pd.read_sql_query(sql, engine)

vendor  total     qty         sum
0  Pernod Ricard USA/Austin Nichols  42338  403450  7880365.88

**Prompt 5 Answer**:   
Vendor: Pernod Ricard Total: 42338 QTY: 403450 Sum: 7880365.88

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Create a column called `Markup` that shows the difference between the `btl_price` and the `state_btl_cost`. Take a screenshot and paste it into your notebook.

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [14]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT (btl_price - state_btl_cost) as Markup
FROM sales
LIMIT 5
"""

pd.read_sql_query(sql, engine)

markup
0  $5.75
1  $5.75
2  $5.75
3  $8.58
4  $2.50

**Prompt 6 Answer**: <img src="./images/Sshot1.png"/>

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

> Note: To make this an easier question, you may take the simple average of percentage markups across all transactions for a store. You do **not** need to weight the average markup by the number of bottles purchased per transaction. If you want to practice your SQL skills, though, consider this a bonus challenge!

In [15]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT store, avg(((btl_price - state_btl_cost)/state_btl_cost)) AS percentage_markup
FROM sales
GROUP BY store
ORDER BY percentage_markup DESC
LIMIT 5
"""

pd.read_sql_query(sql, engine)

store  percentage_markup
0   4013           0.525471
1   4024           0.521499
2   4266           0.518903
3   3833           0.515094
4   3660           0.514591

**Prompt 7 Answer**: Store 4013

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [20]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT sales.store, 
    avg(((btl_price - state_btl_cost)/state_btl_cost)) AS percentage_markup, 
    min((btl_price - state_btl_cost)/state_btl_cost) as min, 
    max((btl_price - state_btl_cost)/state_btl_cost) as max, 
    s.name, 
    s.store_address
FROM sales
INNER JOIN stores AS s ON sales.store = s.store
GROUP BY sales.store, s.name, s.store_address
ORDER BY percentage_markup DESC
LIMIT 5
"""

pd.read_sql_query(sql, engine)

store  percentage_markup       min       max                        name  \
0   4013           0.525471  0.498571  6.708763      Wal-Mart 0841 / Tipton   
1   4024           0.521499  0.498462  6.708763  Wal-Mart 1546 / Iowa Falls   
2   4266           0.518903  0.498462  6.708763  Wal-Mart 1683 / Shenandoah   
3   3833           0.515094  0.498462  6.708763    Wal-Mart 3394 / Atlantic   
4   3660           0.514591  0.498462  6.708763   Wal-Mart 2935 / Knoxville   

                                       store_address  
0  1126 Highway 38 North\nTipton, IA 527720000\n(...  
1  840 S Oak\nIowa Falls, IA 501260000\n(42.50295...  
2  705 S Fremont\nShenandoah, IA 516010000\n(40.7...  
3  1905 East 7th St\nAtlantic, IA 500220000\n(41....  
4  814 W Bell Ave\nKnoxville, IA 501380000\n(41.3...

**Prompt 8 Answer**:  
1. Wal-Mart 0841 / Tipton, IA  
2. Wal-Mart 1546 / Iowa Falls, IA  
3. Wal-Mart 1683 / Shenandoah, IA  
4. Wal-Mart 3394 / Atlantic, IA  
5. Wal-Mart 2935 / Knoxville  

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery). Otherwise, keep an eye out for the solution key!

In [23]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT sales.county, sum(sales.bottle_qty)/counties.population as bottlespercapita, round(avg(sales.liter_size), 1) as bottlesize
FROM sales
INNER JOIN counties ON counties.county = sales.county
GROUP BY sales.county, counties.population
ORDER BY bottlespercapita DESC
LIMIT 9
"""

pd.read_sql_query(sql, engine)

county  bottlespercapita  bottlesize
0      Dickinson                22       974.2
1     Black Hawk                15       861.0
2           Polk                15       891.9
3    Cerro Gordo                14       986.4
4          Scott                13       869.9
5        Johnson                13       895.2
6           Linn                12       885.1
7        Kossuth                11      1040.5
8  Pottawattamie                11       890.9

**Prompt 9 Answer**:
1. Dickinson  
2. Black Hawk  
3. Polk  
4. Cerro Gordo  
5. Scott  
6. Johnson    
7. Linn  
8. Kossuth  
9. Pottawattamie  

Kossuth sells the largest bottle size.

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [24]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT count(store)
FROM sales
WHERE (btl_price * bottle_qty) != cast(total AS money)
"""

pd.read_sql_query(sql, engine)

pd.read_sql_query(sql, engine)

count
0      0

**Prompt 10 Answer**: